clients/src/main/java/org/apache/kafka/common/config/provider/EnvVarConfigProvider.java

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.config.provider;
+
+import org.apache.kafka.common.config.ConfigData;
+import org.apache.kafka.common.config.ConfigException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class EnvVarConfigProvider implements ConfigProvider {
+
+    public static final String ALLOWLIST_PATTERN_CONFIG = "allowlist.pattern";
+    public static final String ALLOWLIST_PATTERN_CONFIG_DOC = "A pattern / regular expression that needs to match for environment variables" +
+            " to be used by this config provider.";
+    private final Map<String, String> envVarMap;
+    private Map<String, String> filteredEnvVarMap;
+    private Pattern envVarPattern;
+
+    public EnvVarConfigProvider() {
+        envVarMap = getEnvVars();
+    }
+
+    public EnvVarConfigProvider(Map<String, String> envVarsAsArgument) {
+        envVarMap = envVarsAsArgument;
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(EnvVarConfigProvider.class);
+
+    @Override
+    public void configure(Map<String, ?> configs) {
+        if (configs.containsKey(ALLOWLIST_PATTERN_CONFIG)) {
+            envVarPattern = Pattern.compile(
+                    String.valueOf(configs.get(ALLOWLIST_PATTERN_CONFIG))
+            );
+        } else {
+            envVarPattern = Pattern.compile(".*");
+            log.info("No pattern for environment variables provided. Using default pattern '(.*)'.");
+        }
+
+        filteredEnvVarMap = envVarMap.entrySet().stream()
+                .filter(envVar -> envVarPattern.matcher(envVar.getKey()).matches())
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)
+                );
+    }
+
+    @Override
+    public void close() throws IOException {
+    }
+
+    /**
+     * @param s unused
+     * @return returns environment variables as configuration
+     */
+    @Override
+    public ConfigData get(String s) {
+        return get(s, null);
+    }
+
+    /**
+     * @param path    path, not used for environment variables
+     * @param keys the keys whose values will be retrieved.
+     * @return the configuration data.
+     */
+    @Override
+    public ConfigData get(String path, Set<String> keys) {
+
+        if (path != null && !path.isEmpty()) {
+            log.error("Path is not supported for EnvVarConfigProvider, invalid value '{}'", path);
+            throw new ConfigException("Path is not supported for EnvVarConfigProvider, invalid value '" + path + "'");
+        }
+
+        if (envVarMap == null) {
```

Comment:
Do we need this check? If `envVarMap` is `null` we will have hit a NPE in `configure()` already

Reply:
True - needed to double-check but you are right!